File: 0095 Unique Binary Search Trees II/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if (!n) return vector<TreeNode*>();
        return dfs(1, n);
    }

    vector<TreeNode*> dfs(int l, int r)
    {
        vector<TreeNode*>res;
        if (l > r)
        {
            res.push_back(NULL);
            return res;
        }
        for (int i = l; i <= r; i ++ )
        {
            vector<TreeNode*>left = dfs(l, i - 1), right = dfs(i + 1, r);
            for (auto &lc : left)
                for (auto &rc : right)
                {
                    TreeNode *root = new TreeNode(i);
                    root->left = lc;
                    root->right = rc;
                    res.push_back(root);
                }
        }
        return res;
    }
};

void print_bfs(TreeNode* p)
{
    if (!p) return;
    std::vector<std::string> retv;
    std::queue<TreeNode *> q;
    q.push(p);
    do {
        TreeNode *node = q.front(); q.pop();
        if (node) 
            retv.push_back(std::to_string(node->val));
        else {
            retv.push_back("#");
            continue;
        }
        q.push(node->left); 
        q.push(node->right);
    } while (!q.empty());
    
    auto found = std::find_if(retv.rbegin(), retv.rend(), [](const std::string &s){ return s != "#"; });
    retv.erase(found.base(), retv.end());
    for (const auto &s : retv)
        std::cout << s << ",";
    std::cout << "\b " << std::endl; 
}

int main()
{
    Solution s;
    for (auto p : s.generateTrees(3))
        print_bfs(p);
} 

File: 0095 Unique Binary Search Trees II/solution.py
#! python3

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        if(n == 0):    return []
        
        return self.dfs(1, n)
    
    def dfs(self, s, e):
        if s > e:
            return [None]
        
        res = []
        for cur_root in range(s, e + 1):
            left = self.dfs(s, cur_root - 1)
            right = self.dfs(cur_root + 1, e)
            
            for l in left:
                for r in right:
                    root = TreeNode(cur_root)
                    root.left = l
                    root.right= r
                    res.append(root)
        return res

def print_bfs(p):
    if not p:  return
    res, q = [], []
    q.append(p)
    while len(q):
        node = q[0]
        q.pop(0)
        if node:
            res.append(str(node.val))
        else:
            res.append('#')
            continue
        q.append(node.left)
        q.append(node.right)
    for _ in res[::-1]:
        if _ == '#':
            res.pop()
        else:
            break
    print(','.join(res))


if __name__ == "__main__":
    for p in Solution().generateTrees(3):
        print_bfs(p)
